cosmos/precompile/erc20/helpers.go

```diff
@@ -0,0 +1,159 @@
+var salt uint64
+
+// deployPolarisERC20Contract deploys a new ERC20 token contract by calling back into the EVM.
+func (c *Contract) deployPolarisERC20Contract(
+	ctx sdk.Context,
+	evm ethprecompile.EVM,
+	deployer common.Address,
+	name string,
+	endowment *big.Int,
+) (common.Address, error) {
+	plugin := c.GetPlugin()
+	plugin.EnableReentrancy(ctx)
+	defer plugin.DisableReentrancy(ctx)
+
+	// deploy new ERC20 token contract
+	code := common.FromHex(pbindings.PolarisERC20MetaData.Bin)
+	args, err := c.polarisERC20ABI.Pack("", name, name)
+	if err != nil {
+		return common.Address{}, err
+	}
+	suppliedGas := ctx.GasMeter().GasRemaining()
+	_, contractAddr, gasRemaining, err := evm.Create2(
+		vm.AccountRef(deployer), append(code, args...), suppliedGas, endowment, uint256.NewInt(salt),
+	)
+	salt++
+
+	// consume gas used by EVM during ERC20 deployment
+	ctx.GasMeter().ConsumeGas(suppliedGas-gasRemaining, "Polaris ERC20 deployment")
+	return contractAddr, err
+}
+
+// callPolarisERC20Mint mints ERC20 tokens by calling back into the EVM.
+func (c *Contract) callPolarisERC20Mint(
+	ctx sdk.Context,
+	evm ethprecompile.EVM,
+	caller common.Address,
+	token common.Address,
+	to common.Address,
+	amount *big.Int,
+) error {
+	plugin := c.GetPlugin()
+	plugin.EnableReentrancy(ctx)
+	defer plugin.DisableReentrancy(ctx)
+
+	// call ERC20 contract to mint
+	input, err := c.polarisERC20ABI.Pack("mint", to, amount)
+	if err != nil {
+		return err
+	}
+	suppliedGas := ctx.GasMeter().GasRemaining()
+	_, gasRemaining, err := evm.Call(
+		vm.AccountRef(caller), token, input, suppliedGas, big.NewInt(0),
+	)
+
+	// consume gas used by EVM during ERC20 mint
+	ctx.GasMeter().ConsumeGas(suppliedGas-gasRemaining, "Polaris ERC20 mint")
+	return err
+}
+
+// callERC20TransferFrom transfers ERC20 tokens by calling back into the EVM.
+func (c *Contract) callERC20TransferFrom(
+	ctx sdk.Context,
+	evm ethprecompile.EVM,
+	caller common.Address,
+	token common.Address,
+	from common.Address,
+	to common.Address,
+	amount *big.Int,
+) error {
+	plugin := c.GetPlugin()
+	plugin.EnableReentrancy(ctx)
+	defer plugin.DisableReentrancy(ctx)
+
+	// call ERC20 contract to transferFrom
+	input, err := c.polarisERC20ABI.Pack("transferFrom", from, to, amount)
+	if err != nil {
+		return err
+	}
+	suppliedGas := ctx.GasMeter().GasRemaining()
+	_, gasRemaining, err := evm.Call(
+		vm.AccountRef(caller), token, input, suppliedGas, big.NewInt(0),
+	)
+
+	// consume gas used by EVM during ERC20 transferFrom
+	ctx.GasMeter().ConsumeGas(suppliedGas-gasRemaining, "ERC20 transferFrom")
+	return err
+}
+
+func (c *Contract) callERC20Approve(
+	ctx sdk.Context,
+	evm ethprecompile.EVM,
+	caller common.Address,
+	token common.Address,
+	spender common.Address,
+	amount *big.Int,
+) error {
+	plugin := c.GetPlugin()
+	plugin.EnableReentrancy(ctx)
+	defer plugin.DisableReentrancy(ctx)
+
+	// call ERC20 contract to approve
+	input, err := c.polarisERC20ABI.Pack("approve", spender, amount)
```

Comment:
you want the user to have to call approve first. We shouldn't be caling approve on behalf of the user, thats kinda cooked.

Reply:
it does feel weird but its basically part of the transferFrom. ERC20 token owner approves the "caller" to spend, but usually owner == caller. In the other way, the erc20 module contract has to approve "caller" to spend, because caller is transferring from module contract to owner.